File: app/src/main/Java/com/vaios/holobar/TextToAudio.java
package com.vaios.holobar;

import android.content.Context;
import android.util.Log;
import java.io.IOException;
import ai.onnxruntime.OrtException;

public class TextToAudio {
    private static final String TAG = "TExtToAudio";
    private static final String VITS_MODEL_FILE = "vits_model.onnx";
    private static final String PHONEMIZER_MODEL_FILE = "phonemizer_model.onnx";
    private static final String GEMINI_API_KEY = "AIzaSyBzw4vQsyQFcpA9hr4ncge4BYZW0CcT_ng";

    private VitsOnnxSynthesizer synthesizer;
    private ApiCall apiCall;
    private final Context context;

    // Hardcoded matrix for speaker ID mapping
    private static final int[][] SPEAKER_ID_MATRIX = {
            {0, 79},
            {1, 90},
            {2, 33},
            {3, 109},
            {4, 100}
    };

    public TextToAudio(Context context) {
        this.context = context;
        initializeModels();
    }

    private void initializeModels() {
        Log.d(TAG, "Initializing models");
        try {
            synthesizer = new VitsOnnxSynthesizer(context, VITS_MODEL_FILE);
            Log.d(TAG, "VITS synthesizer initialized");

            Phonemizer.initialize(context.getAssets(), PHONEMIZER_MODEL_FILE);
            Log.d(TAG, "Phonemizer initialized");

            apiCall = new ApiCall(context, GEMINI_API_KEY);
            Log.d(TAG, "ApiCall initialized");

        } catch (IOException | OrtException e) {
            Log.e(TAG, "Error initializing models", e);
            throw new RuntimeException("Error initializing models: " + e.getMessage());
        }
    }

    public void processText(String text, int speakerId, TextToAudioCallback callback) {
        Log.d(TAG, "Processing text. Text length: " + text.length() + ", Speaker ID: " + speakerId);

        apiCall.sendTextToApi(text, speakerId, new ApiCall.ApiCallCallback() {
            @Override
            public void onSuccess(String narration, String status) {
                Log.d(TAG, "API Response received: " + narration);
                Log.d(TAG, "API Status: " + status);

                try {
                    // Map the API speaker ID to the TTS speaker ID
                    int ttsSpeakerId = mapSpeakerId(speakerId);
                    float[] responseAudio = synthesizer.tts(narration, ttsSpeakerId);
                    callback.onSuccess(narration, status, responseAudio);
                } catch (Exception e) {
                    Log.e(TAG, "Error synthesizing audio response", e);
                    callback.onError("Error synthesizing audio: " + e.getMessage());
                }
            }

            @Override
            public void onError(String error) {
                Log.e(TAG, "API call error: " + error);
                callback.onError(error);
            }
        });
    }

    private int mapSpeakerId(int apiSpeakerId) {
        for (int[] mapping : SPEAKER_ID_MATRIX) {
            if (mapping[0] == apiSpeakerId) {
                return mapping[1];
            }
        }
        // Default to the first TTS speaker ID if no mapping is found
        return SPEAKER_ID_MATRIX[0][1];
    }


    // Custom callback interface for TextToAudio
    public interface TextToAudioCallback {
        void onSuccess(String narration, String status, float[] audioData);
        void onError(String error);
    }
}

File: app/src/main/Java/com/vaios/holobar/TextToAudio.kt
package com.vaios.holobar

import android.content.Context
import android.util.Log
import java.io.IOException
import ai.onnxruntime.OrtException

class TextToAudio(private val context: Context) {
    companion object {
        private const val TAG = "TextToAudio"
        private const val VITS_MODEL_FILE = "vits_model.onnx"
        private const val PHONEMIZER_MODEL_FILE = "phonemizer_model.onnx"
        private const val API_KEY_FILE = "gemini_api_key.txt"

        // Hardcoded matrix for speaker ID mapping
        private val SPEAKER_ID_MATRIX = arrayOf(
                intArrayOf(0, 79),
                intArrayOf(1, 90),
                intArrayOf(2, 33),
                intArrayOf(3, 109),
                intArrayOf(4, 100)
        )
    }

    private lateinit var synthesizer: VitsOnnxSynthesizer
    private lateinit var apiCall: ApiCall

    init {
        initializeModels()
    }

    private fun initializeModels() {
        Log.d(TAG, "Initializing models")
        try {
            synthesizer = VitsOnnxSynthesizer(context, VITS_MODEL_FILE)
            Log.d(TAG, "VITS synthesizer initialized")

            Phonemizer.initialize(context.assets, PHONEMIZER_MODEL_FILE)
            Log.d(TAG, "Phonemizer initialized")

            val apiKey = loadApiKey()
            apiCall = ApiCall(context, apiKey)
            Log.d(TAG, "ApiCall initialized")

        } catch (e: IOException) {
            Log.e(TAG, "Error initializing models", e)
            throw RuntimeException("Error initializing models: ${e.message}")
        } catch (e: OrtException) {
            Log.e(TAG, "Error initializing models", e)
            throw RuntimeException("Error initializing models: ${e.message}")
        }
    }

    private fun loadApiKey(): String {
        return try {
            context.assets.open(API_KEY_FILE).bufferedReader().use { it.readText().trim() }
        } catch (e: IOException) {
            Log.e(TAG, "Error loading API key", e)
            throw RuntimeException("Error loading API key: ${e.message}")
        }
    }

    fun processText(text: String, speakerId: Int, callback: TextToAudioCallback) {
        Log.d(TAG, "Processing text. Text length: ${text.length}, Speaker ID: $speakerId")

        apiCall.sendTextToApi(text, speakerId, object : ApiCall.ApiCallCallback {
            override fun onSuccess(narration: String, status: String) {
                Log.d(TAG, "API Response received: $narration")
                Log.d(TAG, "API Status: $status")

                try {
                    // Map the API speaker ID to the TTS speaker ID
                    val ttsSpeakerId = mapSpeakerId(speakerId)
                    val responseAudio = synthesizer.tts(narration, ttsSpeakerId)
                    callback.onSuccess(narration, status, responseAudio)
                } catch (e: Exception) {
                    Log.e(TAG, "Error synthesizing audio response", e)
                    callback.onError("Error synthesizing audio: ${e.message}")
                }
            }

            override fun onError(error: String) {
                Log.e(TAG, "API call error: $error")
                callback.onError(error)
            }
        })
    }

    private fun mapSpeakerId(apiSpeakerId: Int): Int {
        for (mapping in SPEAKER_ID_MATRIX) {
            if (mapping[0] == apiSpeakerId) {
                return mapping[1]
            }
        }
        // Default to the first TTS speaker ID if no mapping is found
        return SPEAKER_ID_MATRIX[0][1]
    }

    interface TextToAudioCallback {
        fun onSuccess(narration: String, status: String, audioData: FloatArray)
        fun onError(error: String)
    }
}